server/src/main/java/org/opensearch/search/profile/query/ConcurrentQueryProfiler.java

```diff
@@ -0,0 +1,124 @@
+    }
+
+    @Override
+    public ContextualProfileBreakdown<QueryTimingType> getQueryBreakdown(Query query) {
+        long threadId = Thread.currentThread().getId();
+        ConcurrentQueryProfileTree profileTree = threadToProfileTree.computeIfAbsent(threadId, k -> new ConcurrentQueryProfileTree());
+        return profileTree.getProfileBreakdown(query);
+    }
+
+    /**
+     * Removes the last (e.g. most recent) element on ConcurrentQueryProfileTree stack.
+     */
+    @Override
+    public void pollLastElement() {
+        long threadId = Thread.currentThread().getId();
+        ConcurrentQueryProfileTree concurrentProfileTree = threadToProfileTree.get(threadId);
+        concurrentProfileTree.pollLast();
+    }
+
+    /**
+     * @return a hierarchical representation of the profiled tree
+     */
+    @Override
+    public List<ProfileResult> getTree() {
+        List<ProfileResult> profileResults = new ArrayList<>();
+        for (Map.Entry<Long, ConcurrentQueryProfileTree> profile : threadToProfileTree.entrySet()) {
+            profileResults.addAll(profile.getValue().getTree());
+        }
+        return profileResults;
+    }
+
+    /**
+     * Begin timing the rewrite phase of a request
+     */
+    @Override
+    public void startRewriteTime() {
+        long threadId = Thread.currentThread().getId();
+        Timer rewriteTimer = new Timer();
+        threadToRewriteTimers.computeIfAbsent(threadId, k -> new LinkedList<>()).add(rewriteTimer);
+        rewriteTimer.start();
+    }
+
+    /**
+     * Stop recording the current rewrite timer
+     */
+    public void stopAndAddRewriteTime() {
+        long threadId = Thread.currentThread().getId();
+        Timer rewriteTimer = threadToRewriteTimers.get(threadId).getLast();
+        rewriteTimer.stop();
+    }
+
+    /**
+     * @return total time taken to rewrite all queries in this concurrent query profiler
+     */
+    @Override
+    public long getRewriteTime() {
+        long totalRewriteTime = 0L;
+        List<Timer> rewriteTimers = new LinkedList<>();
+        threadToRewriteTimers.values().forEach(rewriteTimers::addAll);
+        LinkedList<long[]> mergedIntervals = mergeRewriteTimeIntervals(rewriteTimers);
+        for (long[] interval : mergedIntervals) {
+            totalRewriteTime += interval[1] - interval[0];
+        }
+        return totalRewriteTime;
+    }
+
+    // package private for unit testing
+    LinkedList<long[]> mergeRewriteTimeIntervals(List<Timer> timers) {
+        LinkedList<long[]> mergedIntervals = new LinkedList<>();
+        timers.sort(Comparator.comparingLong(Timer::getEarliestTimerStartTime));
+        for (Timer timer : timers) {
+            long startTime = timer.getEarliestTimerStartTime();
+            long endTime = timer.getEarliestTimerStartTime() + timer.getApproximateTiming();
+            if (mergedIntervals.isEmpty() || mergedIntervals.getLast()[1] < timer.getEarliestTimerStartTime()) {
```

Comment:
nit: `mergedIntervals.getLast()[1] < startTime`

Reply:
Thanks. It's addressed now.